# FCS Lab 3 - MD5, Rainbow Tables

* Name(s): Leong Yun Qin Melody
* Student ID(s): 1004489

## Part I: Hashing Using MD5

* How does the length of the hash correspond to the input string?

The length of the hash is constant no matter the length of the input string. Below are some examples that are tested using the command line tool.

In [1]:
!echo -n "foobar" | md5sum

3858f62230ac3c915f300c664312c63f  -


In [2]:
!echo -n "sutdkillsmysouldrn" | md5sum

53efcae9b8f88a4d87dc158b45ef0cf9  -


In [3]:
# convince you further
!echo -n "ahfniequhfr0e24r52479y5-4uehortwjfdjnd" | md5sum

027e0513a85c4fe55fde0019e4a35118  -


* Are there any visible correlations between the hash and the input string?

No, there are no visible correlations. The random oracle property states so. For two correlated inputs, m1 and m2, the output is completely not correlated

* What are the issues related to the cryptographic weakness of MD5?

Some issues include the fact that using the birthday attack paradox, the probability of a single MD5 collision is 50%. Since the length of the hash output is arbitrarily fixed, any two different messages can produce the same hash value. So, the attacker can violate the preimage resistance cryptographic property.

## Part II: Break Hashes with Brute Force

* How much time did you take in total?

115.49513483047485 seconds

* How much time does it take to crack each string, on average?

In [1]:
#calculate average amount of time to crack each string
print(115.49513483047485/15)

7.69967565536499


7.69967565536499s

* Is it possible to amortize (gradually write off the initial cost of) the brute forcing attempts?

Yes, it is possible to amortize the brute forcing attempts. The birthday attack paradox states that when storing previous brute force attempts, the time complexity is O(n/2) where n is the number of bits.

## Part III: Creating Rainbow Tables

* Is rcrack faster/slower than your script ex2.py? By how much faster/slower is it?

In [2]:
#rcrack is faster by how much
print(115.49513483047485 - 6.74)

108.75513483047486


rcrack is faster. Rcrack is faster by 108.75513483047486 seconds

* Do you observe any advantages or disadvantages of using rainbowcrack?

Advantage: Computation of hashes are fast. Disadvantage: more space neeeded to compute the hashes.

## Part IV: Salt

* What is the observed differences between your ease of cracking the salted vs the unsalted plaintexts?

The salted plaintext took longer to crack. The salted plaintext took 6.94 seconds as compared to 6.74 seconds for the unsalted text

* Report the difference in time observed to crack.

0.20 seconds

* Explain any differences between salted and non salted rcrack strategies.

The salted rcrack requires the rainbow tables to be regenerated. 

## Part V: Hash Breaking Competition

* What is the approach you used to crack the hashes

I used the rainbow table attack approach. Pre compute and check for precomputed hashes. Then use common words, take the MD5 hash and then check

* How you decided or designed your approach

First, I use precomputed existing lookup tables to compute the MD5 password hashing. There are a list of common words without salt. There is a lookup table for such common passwords. We webscrape the words from wikipedia to form the word database. I also use brute force hybrid on the hashes.txt to make the decryption more effective.

* Main challenges and limitations of your approach

Challenge:
We noticed as the difficulty level increases, less words can be solved. More often than not, the precomputed lookup tables contain a list of common passwords or phrases that people normally use. Hence, for those hashes that do not have commonly used words, or phrases, the plaintext for these words are unable to be recovered.